src/hyp_map.rs

```diff
@@ -93,6 +97,93 @@ impl HypMapRegion {
     }
 }
 
+/// Represents a virtual address region that must be allocated and populated to be mapped.
+struct HypMapPopulatedRegion {
+    // The address space where this region starts.
+    vaddr: PageAddr<SupervisorVirt>,
+    // Number of bytes of the VA area
+    size: u64,
+    // PTE bits for the mappings.
+    pte_fields: PteFieldBits,
+    // Data to be populated in the VA area
+    data: Vec<u8>,
```

Comment:
Can we avoid this extra `memcpy()` from the ELF to this temporary buffer?

Reply:
You're describing my Thursday night. Originally I had kept the reference from `'elf` to here. It is sane because afterall the elf is `'static`. But turns out that at least I couldn't understand how to let Rust understand in a non convoluted way that there were to lifetimes here, one of the underlying data structure and the other of the `user_map`.

In general the structure needs to have a life time, the code becomes much much harder to follow and should essentially hold the ElfMap in it to keep the reference lifetime valid. Just over complication for a language that clearly doesn't like these sort of things.

This would also remove generality to the whole concept of a populated region.

In general, much more complicated code, to save a few kbs of RAM.

Reply:
I think you had your lifetimes backwards in `ElfMap::segments()`. With that fixed, this becomes pretty straightforward, see https://github.com/abrestic-rivos/salus/tree/for-glg/user_mode

Reply:
Whops. That's a few hours of my life I'll never get back. Thank you!

Reply:
Changed to your code. Thanks again.